core/kernel/user_ta.c

```diff
@@ -163,9 +165,15 @@ static TEE_Result user_ta_enter(struct ts_session *session,
 	else
 		memset(usr_params, 0, sizeof(*usr_params));
 
+#ifdef CFG_TA_PAUTH_ENABLE
+	key_hi = utc->ta_ctx.ts_ctx.pauth_key_hi;
+	key_lo = utc->ta_ctx.ts_ctx.pauth_key_lo;
+#endif
+
 	res = thread_enter_user_mode(func, kaddr_to_uref(session),
 				     (vaddr_t)usr_params, cmd, usr_stack,
 				     utc->uctx.entry_func, utc->uctx.is_32bit,
+				     key_hi, key_lo,
```

Comment:
We're accumulating quite a bit of arguments for `thread_enter_user_mode()`. How about replacing everything after the `entry_func` parameter with `utc` and let `thread_enter_user_mode()` dig out the rest of the parameters there? In the process we'll get rid of a bunch of `#ifdef CFG_TA_PAUTH_ENABLE`.

Reply:
+1

Reply:
thread_enter_user_mode() is also called from core/arch/arm/kernel/ldelf_loader.c where utc is not available. I can try passing uctx (i.e  struct user_mode_ctx). It has is_32bit in it and also pointer to ts_ctx from where keys can be derived. But would this work with core/arch/arm/kernel/ldelf_loader.c where when we are calling thread_enter_user_mode(), we use is_arm32 which is globally defined. I think I might be missing something here.

Reply:
Hmm, that was unfortunate. How about passing the two keys as an array where the pointer is retrieved with a helper function that just returns NULL if CFG_TA_PAUTH_ENABLE isn't enabled? That should at least save us one argument and a bunch of ifdefs.

Reply:
I am making some changes to remove the key parameter from this function. Instead we will get teh current session and get the keys from the ts_ctx. Please have a look